tests/phpunit/tests/block-template.php

```diff
@@ -275,6 +275,83 @@ public function test_get_the_block_template_html_does_not_generally_enforce_loop
 		$this->assertSame( array( false, true ), $in_the_loop_logs, 'Main query loop was triggered incorrectly' );
 	}
 
+	/**
+	 * @ticket 58319
+	 *
+	 * @covers ::get_block_theme_folders
+	 * @dataProvider data_get_block_theme_folders
+	 */
+	public function test_get_block_theme_folders( $theme, $expected, $message ) {
+		$this->assertSame( $expected, get_block_theme_folders( $theme ), $message );
+
+		// Test that the result is cached.
+		if ( empty( $theme ) ) {
+			$theme = get_stylesheet();
+		}
+		$cache_key    = 'theme-' . md5( get_theme_root( $theme ) . '/' . $theme );
+		$theme_cache  = wp_cache_get( $cache_key, 'themes' );
+		$cached_value = $theme_cache['block_template_folders'];
```

Comment:
```suggestion
                $wp_theme = wp_get_theme( $theme ):
		$reflection = new ReflectionMethod( $wp_theme, 'cache_get' );
		$reflection->setAccessible( true );
		
		$theme_cache = $reflection->invoke( $wp_theme, 'theme' );
		$cached_value = $theme_cache['block_template_folders'];
```

Call private method like this. 

Another example of this 

https://github.com/WordPress/wordpress-develop/blob/a7c9394b789daad0754f21ca2bd4f6ee4d75b1b7/tests/phpunit/tests/user/queryCache.php#L765-L779

Reply:
Note that `setAccessible( false )` should be called after invocation. This is missing from the example provided, but should have been added there too. I'll follow up on this to correct any similar occurrences in Core.

Reply:
@costdev We should have a util method on the test class to access private properties and methods. It is done a number of times in tests.

Reply:
Thank you @spacedmonkey  for the sample. 
@spacedmonkey  The only thing I am worried here is a possibility of bias getting created when we make a call to `wp_get_theme( $theme )`. If the cache was not set during `get_block_theme_folders` then there is still a chance of it getting set during the call to `wp_get_theme( $theme )`.
Which is why I was I took the long route of creating hash and then check the wp_cache directly.

Again, this is a highly unlikely case as `get_block_theme_folders` itself makes a call to `wp_get_theme( $theme )`. I'll go ahead and modify the test using ReflectionMethod.

Reply:
Couldn't use call this bit before the `get_block_theme_folders`
```
	public function test_get_block_theme_folders( $theme, $expected, $message ) {
                $wp_theme = wp_get_theme( $theme ):
		$reflection = new ReflectionMethod( $wp_theme, 'cache_get' );
		$reflection->setAccessible( true );
                 $wp_theme->cache_delete(); // Clear cache. 
		$this->assertSame( $expected, get_block_theme_folders( $theme ), $message );

		$theme_cache = $reflection->invoke( $wp_theme, 'theme' );
		$cached_value = $theme_cache['block_template_folders'];

		$this->assertSame( $expected, $cached_value, $message . ' to be cached' );
	}
```

Reply:
Perfect @spacedmonkey. It did remove the bias and worked as expected. Thank you.

Reply:
@spacedmonkey I have a similar thought about Reflection helpers. I have a request for feedback out at the moment and based on that, I'll decide whether to open a ticket or whether this should be done on a per-class basis.

For now, the above suggestion is fine with the exception that it's missing `$reflection->setAccessible( false )` after the method is invoked. I'd suggest this:

```php
public function test_get_block_theme_folders( $theme, $expected, $message ) {
	$wp_theme = wp_get_theme( $theme ):

	$wp_theme->cache_delete(); // Clear cache. 
	$block_theme_folders = get_block_theme_folders( $theme );

	$reflection = new ReflectionMethod( $wp_theme, 'cache_get' );
	$reflection->setAccessible( true );
	$theme_cache = $reflection->invoke( $wp_theme, 'theme' );
	$reflection->setAccessible( false );

	$cached_value = $theme_cache['block_template_folders'];

	$this->assertSame( $expected, $block_theme_folders, $message );
	$this->assertSame( $expected, $cached_value, $message . ' to be cached' );
}
```

This way, accessibility is only granted for as long as necessary, and it's reset before the assertions in case either of them fails.